crates/sparrow-sources/src/in_memory.rs

```diff
@@ -0,0 +1,221 @@
+use std::sync::{Arc, RwLock};
+
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use error_stack::{IntoReportCompat, ResultExt};
+use futures::{Stream, StreamExt, TryStreamExt};
+
+use sparrow_batch::Batch;
+use sparrow_core::ReadConfig;
+use sparrow_interfaces::Source;
+use sparrow_merge::old::homogeneous_merge;
+
+use crate::Error;
+
+/// Wraps the mutable in-memory batch container.
+pub struct InMemory {
+    /// The prepared schema.
+    ///
+    /// Note this is not the `projected_schema`, which is the schema
+    /// after applying column projections.
+    prepared_schema: SchemaRef,
+    /// The in-memory batches.
+    data: Arc<InMemoryBatches>,
+}
+
+impl InMemory {
+    pub fn new(queryable: bool, schema: SchemaRef) -> error_stack::Result<Self, Error> {
+        let data = Arc::new(InMemoryBatches::new(queryable, schema.clone()));
+        let source = Self {
+            prepared_schema: schema,
+            data,
+        };
+        Ok(source)
+    }
+
+    /// Add a batch, publishing it to the subscribers.
+    pub async fn add_batch(&self, batch: RecordBatch) -> error_stack::Result<(), Error> {
+        self.data.add_batch(batch).await
+    }
+}
+
+impl Source for InMemory {
+    fn schema(&self) -> SchemaRef {
+        self.prepared_schema.clone()
+    }
+
+    fn read(
+        &self,
+        read_config: ReadConfig,
+    ) -> futures::stream::BoxStream<'_, error_stack::Result<Batch, sparrow_interfaces::Error>> {
+        // TODO: Do we get any benefit from projecting the schema here?
+        // Or should the prepared schema just match the projected for in-memory sources?
```

Comment:
I think we don't need to do field projection here, since we already have the batch in memory and need to read it as a `RecordBatch` anyway -- we can do a `ProjectStep` to get the fields we want afterwards. No need to overload the read by messing with columns/schemas, since (I don't think) there's a significant performance benefit. 

(For parquet files that can read per column it's different and we should do field projection as we read)

Reply:
What I worry about with that is it seems to lead to compilation needing to know if `source.supports_projection()` or something like that. So it may be beneficial to push it down to the read in all cases, so it can implement it as appropriate.

Reply:
I think it's good to defer / push to a follow-up.

1. We don't do projection pushdown so it will never need to do it.
2. We're deferring projection pushdown.

So, would suggest (a) adding "implement projection pushdown" to the deferred list wherever we're tracking that and (b) maybe passing the `projected_datatype: &DataType` into the read method (which we'll need to do pushdown), and then you can just do a `assert_eq!(&DataType::Struct(schema.fields.clone()), projected_datatype, "Projection not yet supported");` as the minimum fix here (so we get a runtime assertion to debug)